### PR TITLE
Improvements around services with API keys

### PR DESCRIPTION
**Pull request by Scrumplex**

### CODE REVIEW

File: launcher/ui/pages/global/APIPage.cpp [hunk 1]
```diff
@@ -36,12 +36,16 @@
  */
 
 #include "APIPage.h"
+#include <quuid.h>
+#include <qvalidator.h>
```
Comment (LennyMcLennington) [inline launcher/ui/pages/global/APIPage.cpp]: Why is this including the internal Qt headers? This does not seem like a good idea, as far as I can see those are an implementation detail that could be changed at any time.

**Review by LennyMcLennington — DISMISSED** [review]

**Comment by LennyMcLennington** [comment]
Also what's the purpose of renaming it to flame

File: launcher/ui/pages/global/APIPage.cpp (hunk 1 above)
Comment (Scrumplex) [reply launcher/ui/pages/global/APIPage.cpp]: That's just clangd or nvim-lspconfig being stupid.

**Comment by Scrumplex** [comment]
> Also what's the purpose of renaming it to flame

It's called Flame everywhere except for the user-facing parts. Just want to keep it consistent.

**Comment by LennyMcLennington** [comment]
> > Also what's the purpose of renaming it to flame
> 
> It's called Flame everywhere except for the user-facing parts. Just want to keep it consistent.

I think the description of the option in the cmake script counts as user facing but the rest seems fine to me.

File: launcher/ui/pages/global/APIPage.cpp [hunk 2]
```diff
@@ -137,8 +145,8 @@ void APIPage::loadSettings()
     ui->msaClientID->setText(msaClientID);
     QString metaURL = s->get("MetaURLOverride").toString();
     ui->metaURL->setText(metaURL);
-    QString curseKey = s->get("CFKeyOverride").toString();
-    ui->curseKey->setText(curseKey);
+    QString flameKey = s->get("FlameKeyOverride").toString();
```
Comment (DioEgizio) [inline launcher/ui/pages/global/APIPage.cpp:148]: Wouldn't changing this remove the current override for people who already have overrides?

Comment (Scrumplex) [reply launcher/ui/pages/global/APIPage.cpp:148]: Migration code: https://github.com/PolyMC/PolyMC/blob/fb10a90fe3f1c730af7e2cd7905a3f154407ab85/launcher/Application.cpp#L711-L721

Comment (DioEgizio) [reply launcher/ui/pages/global/APIPage.cpp:148]: oh nvm

File: launcher/net/Download.cpp [hunk 3]
```diff
@@ -117,8 +118,9 @@ void Download::executeTask()
     }
 
     request.setHeader(QNetworkRequest::UserAgentHeader, BuildConfig.USER_AGENT);
-    if (request.url().host().contains("api.curseforge.com")) {
-        request.setRawHeader("x-api-key", APPLICATION->getCurseKey().toUtf8());
+    if (APPLICATION->currentCapabilities() & Application::SupportsFlame
```
Comment (flowln) [inline launcher/net/Download.cpp:121]:
```suggestion
    if ((APPLICATION->currentCapabilities() & Application::SupportsFlame)
```

File: launcher/ui/pages/global/AccountListPage.cpp [hunk 4]
```diff
@@ -94,7 +94,7 @@ AccountListPage::AccountListPage(QWidget *parent)
     updateButtonStates();
 
     // Xbox authentication won't work without a client identifier, so disable the button if it is missing
-    if (APPLICATION->getMSAClientID().isEmpty()) {
+    if (APPLICATION->currentCapabilities() & Application::SupportsMSA) {
```
Comment (flowln) [inline launcher/ui/pages/global/AccountListPage.cpp]: Shouldn't this be the opposite? If it has support for MSA, it shouldn't go into this if, right?

**Review by flowln — COMMENTED** [review]
You really liked those Qt flag thingies didn't you? xD

**Comment by Scrumplex** [comment]
> You really liked those Qt flag thingies didn't you? xD

Yes I do! :D

**Review by flowln — APPROVED** [review]
Only a couple things:
- Where did you get the FlameAPI key regex from? Will it always be like that for sure?
- Maybe we could connect the `QValidator`s `changed` signal to show somewhere in the UI whether the key is valid / invalid
- It's probably a good idea to rebase onto #708 (or the other way around), since it messes with the API page as well

looks good otherwise :D

**Comment by Scrumplex** [comment]
> Where did you get the FlameAPI key regex from? Will it always be like that for sure?

CurseForge uses bcrypt2 hashes as for all api keys. So this is just a regex for bcrypt2 hashes

> Maybe we could connect the QValidators changed signal to show somewhere in the UI whether the key is valid / invalid

Validators prevent you from entering text that doesn't match

> It's probably a good idea to rebase onto https://github.com/PolyMC/PolyMC/pull/708 (or the other way around), since it messes with the API page as well

sure

**Comment by flowln** [comment]
> Validators prevent you from entering text that doesn't match

Yeah, but it will not tell you if you entered an incomplete key, for instance (neither will it tell you why you can't type a character, but that's a `QValidator` issue, not ours :p). Would be cool if we had an indication that the entered key is valid (read complete) or not, instead of just blocking user input and not saying anything.

**Comment by Scrumplex** [comment]
Right now settings does not allow us to block saving if something is invalid. This might be possible to do by creating a dialog that is only acceptable if the input is valid

**Comment by flowln** [comment]
> Right now settings does not allow us to block saving if something is invalid. This might be possible to do by creating a dialog that is only acceptable if the input is valid

I don't think we need to disallow saving an incomplete key, it's just about having some visual indication that something is missing to the user, just a small UX thing

**Comment by DioEgizio** [comment]
conflicting with develop

**Review by ZekeZDev — APPROVED** [review]

**Comment by DioEgizio** [comment]
Conflicting with develop

**Review by txtsd — APPROVED** [review]

**Pull request merged**